### DISCUSSION TITLE
Rustdesk on FreeBSD

### DISCUSSION THREAD

**Discussion by edschofield15** [rustdesk/rustdesk#5466] [Q&A]
I'm trying to migrate my laptop over to FreeBSD from Linux. Most of the applications that I use are already available on FBSD. However it doesn't seem like Rustdesk is. Has anyone had any luck with building/porting the client over to BSD that could give me some pointers?

**Comment by SergeyMy** [comment]
self-assembly only.
https://github.com/rustdesk/rustdesk#install-vcpkg
you will need:
1. rust
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
2. Install vcpkg
3. probably LLVM

PS as far as I know BSD conflicts with the GPL somewhat - there may be nuances :)

**Comment by DevJupyHUB** [comment]
@edschofield15 "building/porting **the client** over to BSD"

ping @madpilot78 - he maintains the server, maybe he can help with the client :)
https://www.freshports.org/net/rustdesk-server/

BTW:
others would also like to see it as a BSD port(s)

**Comment by madpilot78** [comment]
Hi!

I actually have porting the rustdesk client to FreeBSD, since it would be useful to me too, at least to connect to remote machines (sharing the local one is less of interest to me).

I've had little time to investigate this but at present the biggest roadblock (based on the little investigation I have done) is that the build instructions depend on Sciter, which is closed source AND no binaries are provided for FreeBSD [1].

The README also states that Flutter can be user, but the procedure is not documented. If any rustdesk developer could share some insight on how to build rustdesk using flutter that would help. But please note that there is no official flutter port for FreeBSD, so, before starting work on a rustdesk port, I would need to create a flutter port. This could take some time.

[1] the fact that sciter is closed source is in itself not a problem, the FreeBSD ports collection does include binary only closed source software. But if upstream is not providing FreeBSD binaries, there is no way to create a port.

Related to this, I'd like to reply to @SergeyMy noting that there is no practical conflict in FreeBSD about porting GNU software as external port. The GPL, nor the BSD license have no rule forbidding users to download sources compile the themselves and using them on their system, which is what the ports collection does. There could be a problem including GPL v3 software in the FreeBSD "base" distribution, but this is not what we are talking about.

**Reply by edschofield15** [reply]
This is basically my use case. Connecting to remote machines rather than providing access to the local one. Interested to see where this goes. 👍

**Comment by SergeyMy** [comment]
developers abandoned Sciter for more than a year :), the first branch is Flutter! precisely because of the availability of open source code.
assembly is not as difficult as it seems, although there are no text instructions, but ....

1.rust
    curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs/ | sh
2.    Install vcpkg
3.    probably LLVM
4. wget flutter and add PATH (wget https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_3.10.6-stable.tar.xz)
5. get and local install flutter_rust_bridge (just at this stage, the correctness of the work of RUST, LLVM, FLUTTER will be checked)
6. build

PS
instructions in the form of package code in:
**rustdesk/.github/workflows** (here is the most important and most relevant along with the versions)

**Reply by madpilot78** [reply]
LLVM is already present in FreeBSD base (tailored for it) or as a port (more complete version), and rust is also available as a port, so those are easy.

vcpkg (which I know nothing about) is not available for FreeBSD easily, same goes for Flutter; so before porting rustdesk client, those two need to be ported, which will take tame, since I can devote only a minor portion of my time to this endeavor.

Anyway I'm going to take a look.

**Comment by DevJupyHUB** [comment]
@madpilot78  "vcpkg (which I know nothing about) is not available for FreeBSD easily"

oh yes it will be a long and bumpy road :)
https://github.com/microsoft/vcpkg/pull/4917

**Reply by SergeyMy** [reply]
the post is outdated, on the version that is currently being assembled by rustdesk there is a mention of FreBSD
![изображение](https://github.com/rustdesk/rustdesk/assets/131688106/f5422f31-a150-416a-8862-e8a506f0ae8c)
but the way is not easy :)

**Reply by madpilot78** [reply]
That's a good thing, Hoping it does work as advertised, will require testing.

**Comment by DevJupyHUB** [comment]
@SergeyMy 
Yupp, one step closer :)

**Comment by SergeyMy** [comment]
don't forget :) flutter //added above

**Comment by 4mig4** [comment]
I see that we are in 2023 and still no rustdesk client in the ports/pkgs in FreeBSD

**Comment by SergeyMy** [comment]
FreeBSD is not very GPL friendly

**Comment by madpilot78** [comment]
@4mig4 @SergeyMy 

> FreeBSD is not very GPL friendly

This has nothing to do with GPL The ports collection is full of GPL software. Any kind of useful GPL software can be added to the ports tree without any friction. (FreeBSD base is a slightly different matter for very specific reasons, which I will not discuss, because we are not talking about FreeBSD base in this case, and I also I'm not that well informed about that subject)

Creating a rustdesk port is a complex project. Lots of required parts are missing FreeBSD ports collection that would need to be ported even before one can start looking at porting rustdesk itself.

Since I did state I'd like to create a port, I feel like this kind of critics is somewhat leveled at me also. But I simply do not have time and resources right now to dedicate for this.

I'll state that again: I'd like to make a port, but need to find the time to do that, maybe some day in the future. No deadlines. No estimates. No promises.

In open source (GPL, BSD, MIT whatever) things happen is someone actually takes the time and efforts to work on them [1], pushing others rarely works...if ever.

So if you have time and resources you can start your own porting effort, nobody will stop you. Even if you are unable to produce a working port in the end, you will learn something from the endeavor.

And keep in mind that most work done on the FreeBSD ports tree is performed by volunteers on their own free time, and using heir own technical resources, so it is unfair to push deadlines and expectations on them.

Said that I don't have anything to add to this thread until the day (sooner, later or never whatever that will be) I will have some code/patch to show regarding this effort.

[1] Work as actually sitting down as in writing code, do testing and so on.

**Reply by DevJupyHUB** [reply]
@madpilot78 "_And keep in mind that most work done on the FreeBSD ports tree is performed by volunteers on their own free time..."_

I followed it all the way, and I'm still following it today (because I use it) on pfSense from v1.1 - that's why I agree with you :)

- although there is now a serious team working on this, Netgate..., but always have to start somewhere....

I would definitely like to see your work, as your time permits, THX

**Comment by tschettervictor** [comment]
Just curious as to the status of a FreeBSD client. Being able to connect to GhostBSD would be great.

**Reply by tschettervictor** [reply]
Sigh...
